crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -55,38 +55,61 @@ pub(super) struct ScopedDefinitionId;
 #[newtype_index]
 pub(super) struct ScopedConstraintId;
 
+/// A newtype-index for a [`crate::semantic_index::branching_condition::BranchingCondition`] in a particular scope.
+#[newtype_index]
+pub(super) struct ScopedBranchingConditionId;
+
 /// Can reference this * 64 total definitions inline; more will fall back to the heap.
 const INLINE_BINDING_BLOCKS: usize = 3;
 
 /// A [`BitSet`] of [`ScopedDefinitionId`], representing live bindings of a symbol in a scope.
 type Bindings = BitSet<INLINE_BINDING_BLOCKS>;
-type BindingsIterator<'a> = BitSetIterator<'a, INLINE_BINDING_BLOCKS>;
+type ReverseBindingsIterator<'a> = ReverseBitSetIterator<'a, INLINE_BINDING_BLOCKS>;
 
 /// Can reference this * 64 total declarations inline; more will fall back to the heap.
 const INLINE_DECLARATION_BLOCKS: usize = 3;
 
 /// A [`BitSet`] of [`ScopedDefinitionId`], representing live declarations of a symbol in a scope.
 type Declarations = BitSet<INLINE_DECLARATION_BLOCKS>;
-type DeclarationsIterator<'a> = BitSetIterator<'a, INLINE_DECLARATION_BLOCKS>;
+type ReverseDeclarationsIterator<'a> = ReverseBitSetIterator<'a, INLINE_DECLARATION_BLOCKS>;
 
 /// Can reference this * 64 total constraints inline; more will fall back to the heap.
 const INLINE_CONSTRAINT_BLOCKS: usize = 2;
 
 /// Can keep inline this many live bindings per symbol at a given time; more will go to heap.
 const INLINE_BINDINGS_PER_SYMBOL: usize = 4;
 
-/// One [`BitSet`] of applicable [`ScopedConstraintId`] per live binding.
-type InlineConstraintArray = [BitSet<INLINE_CONSTRAINT_BLOCKS>; INLINE_BINDINGS_PER_SYMBOL];
-type Constraints = SmallVec<InlineConstraintArray>;
-type ConstraintsIterator<'a> = std::slice::Iter<'a, BitSet<INLINE_CONSTRAINT_BLOCKS>>;
+/// Which constraints apply to a given binding?
+type Constraints = BitSet<INLINE_CONSTRAINT_BLOCKS>;
+
+type InlineConstraintArray = [Constraints; INLINE_BINDINGS_PER_SYMBOL];
+
+/// One [`BitSet`] of applicable [`ScopedConstraintId`]s per live binding.
+type ConstraintsPerBinding = SmallVec<InlineConstraintArray>;
+
+/// Iterate over all constraints for a single binding.
+type ConstraintsIterator<'a> = std::slice::Iter<'a, Constraints>;
 type ConstraintsIntoIterator = smallvec::IntoIter<InlineConstraintArray>;
 
+/// Similar to what we have for constraints, but for active branching conditions.
+const INLINE_BRANCHING_BLOCKS: usize = 2;
+const INLINE_BRANCHING_CONDITIONS: usize = 4;
+pub(super) type BranchingConditions = BitSet<INLINE_BRANCHING_BLOCKS>;
+type InlineBranchingConditionsArray = [BranchingConditions; INLINE_BRANCHING_CONDITIONS];
+/// One [`BitSet`] of active [`ScopedBranchingConditionId`]s per live binding.
+type BranchingConditionsPerBinding = SmallVec<InlineBranchingConditionsArray>;
+type BranchingConditionsIterator<'a> = std::slice::Iter<'a, BranchingConditions>;
+type BranchingConditionsIntoIterator = smallvec::IntoIter<InlineBranchingConditionsArray>;
```

Comment:
Do we push a branching condition for every live binding? If we're only pushing an entry for bindings that are known to be always false or true, then I think it might be worth to consider using a `HashMap` instead (or a `Vec` and to linear search) because most code will only have very few "conditions" (the list is not dense).

Reply:
> Do we push a branching condition for every live binding? If we're only pushing an entry for bindings that are known to be always false or true

We don't know that yet while building the semantic index. So we push an empty branching conditions bitset for every binding we encounter.

> it might be worth to consider using a HashMap instead (or a Vec and to linear search) because most code will only have very few "conditions" (the list is not dense).

I guess you are imagining a ScopedDefinitionId => BitSet HashMap?

There are two questions here:
* How many bindings come with branching conditions attached?
* For those that do, out of the set of all possible branching conditions, how many are set for this binding?

Is the concern that there are many bindings without any branching conditions? Or that there are many conditions and bindings usually only refer to a sparse subset?

If it's the former, I'm not sure that is the case. For every given scope, variables that appear directly in that scope would not have any branching conditions attached. But every binding that appears inside a `try` block, any kind of loop, an `if` block etc. would have at least one such branching condition.

Reply:
> If it's the former, I'm not sure that is the case. For every given scope, variables that appear directly in that scope would not have any branching conditions attached. But every binding that appears inside a try block, any kind of loop, an if block etc. would have at least one such branching condition.

Okay, I wasn't aware of that. My assumption was that we'd only push a condition if:

* It's known to be true
* It's known to be false
* An enclosing condition is known to be true or false (to now flag that it's unknown)

I then assumed that this is relatively rare and that a different representation might be faster because it is just "empty" for most scopes.